Issue 575 heap pool memory metrics

kananindzya opened the pull request:
closes #575 
added heap-pool memory metrics

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1228/detail/PR-1228/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1228/detail/PR-1228/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1228/detail/PR-1228/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1228/detail/PR-1228/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #1228 opened]`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-06-10T10:39:00.078+0000

* Duration: 3 min 16 sec


* Commit: e7b499bad85d9e7b49068fdd2c90e840e8e98447




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-06-10T10:42:11.635+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-java/pipelines/apm-agent-java-mbp/pipelines/PR-1228/runs/1/steps/81/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-06-10T10:41:47.177Z] Using shallow fetch with depth 3
[2020-06-10T10:41:47.177Z] Pruning obsolete local branches
[2020-06-10T10:41:47.042Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-10T10:41:47.063Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-06-10T10:41:47.100Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-10T10:41:47.140Z]  > git rev-parse --verify HEAD # timeout=10
[2020-06-10T10:41:47.160Z] No valid HEAD. Skipping the resetting
[2020-06-10T10:41:47.160Z]  > git clean -fdx # timeout=10
[2020-06-10T10:41:47.178Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-06-10T10:41:47.178Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-10T10:41:47.183Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-agent-java.git +refs/pull/1228/head:refs/remotes/origin/PR-1228 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-06-10T10:41:47.905Z] Merging remotes/origin/master commit 42d312bacdff4f272c3e6fc42e032347307bba96 into PR head commit e7b499bad85d9e7b49068fdd2c90e840e8e98447
[2020-06-10T10:41:48.339Z] Merge succeeded, producing 77c28cb666abdae15dbb070d060c70c47d84f2f6
[2020-06-10T10:41:48.339Z] Checking out Revision 77c28cb666abdae15dbb070d060c70c47d84f2f6 (PR-1228)
[2020-06-10T10:41:47.910Z]  > git config core.sparsecheckout # timeout=10
[2020-06-10T10:41:47.919Z]  > git checkout -f e7b499bad85d9e7b49068fdd2c90e840e8e98447 # timeout=15
[2020-06-10T10:41:48.114Z]  > git remote # timeout=10
[2020-06-10T10:41:48.130Z]  > git config --get remote.origin.url # timeout=10
[2020-06-10T10:41:48.150Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-10T10:41:48.156Z]  > git merge 42d312bacdff4f272c3e6fc42e032347307bba96 # timeout=10
[2020-06-10T10:41:48.325Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-06-10T10:41:48.338Z]  > git config core.sparsecheckout # timeout=10
[2020-06-10T10:41:48.350Z]  > git checkout -f 77c28cb666abdae15dbb070d060c70c47d84f2f6 # timeout=15
[2020-06-10T10:41:52.055Z] Commit message: "Merge commit '42d312bacdff4f272c3e6fc42e032347307bba96' into HEAD"
[2020-06-10T10:41:52.055Z] First time build. Skipping changelog.
[2020-06-10T10:41:52.055Z] Cleaning workspace
[2020-06-10T10:41:52.252Z]  > git --version # timeout=10
[2020-06-10T10:41:52.263Z] fatal: bad object 1b4523d4f8cbf97fc5284ea55e1a9af3863f4075
[2020-06-10T10:41:52.054Z]  > git rev-parse --verify HEAD # timeout=10
[2020-06-10T10:41:52.068Z] Resetting working tree
[2020-06-10T10:41:52.069Z]  > git reset --hard # timeout=10
[2020-06-10T10:41:52.204Z]  > git clean -fdx # timeout=10
[2020-06-10T10:41:53.000Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2020-06-10T10:41:53.071Z] Timeout set to expire in 1 hr 0 min
[2020-06-10T10:41:53.102Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2020-06-10T10:41:53.416Z] [INFO] Number of builds to be searched 10
[2020-06-10T10:41:53.683Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-06-10T10:41:53.708Z] Running in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1228/src/github.com/elastic/apm-agent-java
[2020-06-10T10:41:53.748Z] [INFO] gitCheckout: Checkout SCM PR-1228 with some customisation.
[2020-06-10T10:41:53.779Z] [INFO] Override default checkout
[2020-06-10T10:41:53.840Z] Sleeping for 10 sec
[2020-06-10T10:42:04.000Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-06-10T10:42:04.029Z] Cloning the remote Git repository
[2020-06-10T10:42:04.041Z] Cloning repository git@github.com:elastic/apm-agent-java.git
[2020-06-10T10:42:04.065Z]  > git init /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1228/src/github.com/elastic/apm-agent-java # timeout=10
[2020-06-10T10:42:04.085Z] Using reference repository: /var/lib/jenkins/.git-references/apm-agent-java.git
[2020-06-10T10:42:04.086Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-06-10T10:42:04.086Z]  > git --version # timeout=10
[2020-06-10T10:42:04.098Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-10T10:42:04.102Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-06-10T10:42:05.868Z] Checking out Revision e7b499bad85d9e7b49068fdd2c90e840e8e98447 (origin/PR-1228)
[2020-06-10T10:42:05.111Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-10T10:42:05.115Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-06-10T10:42:05.124Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-10T10:42:05.134Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-06-10T10:42:05.134Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-10T10:42:05.137Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-java.git +refs/pull/1228/head:refs/remotes/origin/PR-1228 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2020-06-10T10:42:05.858Z]  > git rev-parse origin/PR-1228^{commit} # timeout=10
[2020-06-10T10:42:05.866Z]  > git config core.sparsecheckout # timeout=10
[2020-06-10T10:42:05.869Z]  > git checkout -f e7b499bad85d9e7b49068fdd2c90e840e8e98447 # timeout=10
[2020-06-10T10:42:06.046Z] Commit message: "deleted non-heap-pool metrics"
[2020-06-10T10:42:06.047Z] First time build. Skipping changelog.
[2020-06-10T10:42:06.060Z]  > git --version # timeout=10
[2020-06-10T10:42:06.078Z] fatal: bad object 1b4523d4f8cbf97fc5284ea55e1a9af3863f4075
[2020-06-10T10:42:06.932Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-06-10T10:42:07.580Z] + git fetch https://****:****@github.com/elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-06-10T10:42:08.583Z] Archiving artifacts
[2020-06-10T10:42:09.277Z] + git rev-parse HEAD
[2020-06-10T10:42:09.656Z] + git rev-parse HEAD
[2020-06-10T10:42:09.975Z] + git rev-parse origin/pr/1228
[2020-06-10T10:42:10.031Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-06-10T10:42:10.269Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-06-10T10:42:11.242Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-java/pulls/1228/reviews return 0 elements
[2020-06-10T10:42:11.319Z] [INFO] githubPrCheckApproved: Title: Issue 575 heap pool memory metrics - User: kananindzya - Author Association: CONTRIBUTOR
[2020-06-10T10:42:11.654Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-06-10T10:42:11.654Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-06-10T10:42:11.744Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2020-06-10T10:42:11.770Z] Sleeping for 5 sec
[2020-06-10T10:42:13.054Z] Stage "Build" skipped due to earlier failure(s)
[2020-06-10T10:42:13.261Z] Stage "Tests" skipped due to earlier failure(s)
[2020-06-10T10:42:13.464Z] Stage "Unit Tests" skipped due to earlier failure(s)
[2020-06-10T10:42:13.476Z] Stage "Smoke Tests 01" skipped due to earlier failure(s)
[2020-06-10T10:42:13.487Z] Stage "Smoke Tests 02" skipped due to earlier failure(s)
[2020-06-10T10:42:13.497Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-06-10T10:42:13.508Z] Stage "Javadoc" skipped due to earlier failure(s)
[2020-06-10T10:42:13.684Z] Failed in branch Unit Tests
[2020-06-10T10:42:13.693Z] Failed in branch Smoke Tests 01
[2020-06-10T10:42:13.702Z] Failed in branch Smoke Tests 02
[2020-06-10T10:42:13.710Z] Failed in branch Benchmarks
[2020-06-10T10:42:13.719Z] Failed in branch Javadoc
[2020-06-10T10:42:13.976Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-06-10T10:42:14.074Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-10T10:42:14.118Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-10T10:42:14.221Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-10T10:42:14.868Z] Running on Jenkins in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1228
[2020-06-10T10:42:15.026Z] [INFO] getVaultSecret: Getting secrets
[2020-06-10T10:42:15.094Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-06-10T10:42:15.878Z] + chmod 755 generate-build-data.sh
[2020-06-10T10:42:15.878Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1228/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1228/runs/1 ABORTED 195531
[2020-06-10T10:42:16.429Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1228/runs/1/steps/?limit=10000 -o steps-info.json

```

</p>
</details>

kananindzya commented on apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java:
```diff
@@ -26,56 +26,91 @@
-            @Override
-            public double get() {
-                return platformMXBean.getHeapMemoryUsage().getMax();
+            if (memoryPoolMXBean.getName().contains("Eden")) {
```
I think it will be more informative. And this detailed information can help to add specific alerts.
Yes, I mean that they will be reported separately but with different lables(I fixed it with last commit)
![image](https://user-images.githubusercontent.com/48118512/84603108-ab9d2800-aead-11ea-9992-a046687a8003.png)

kananindzya commented on apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java:
```diff
@@ -26,56 +26,91 @@
-            public double get() {
-                return platformMXBean.getHeapMemoryUsage().getCommitted();
+
+        addUsedMetric("jvm.memory.heap.used", registry, platformMXBean.getHeapMemoryUsage());
```
fixed

eyalkoren commented on apm-agent-plugins/apm-grpc/apm-grpc-test-latest/src/test/java/co/elastic/apm/agent/grpc/latest/testapp/generated/HelloGrpc.java:
```diff
@@ -42,7 +42,7 @@
  */
 @javax.annotation.Generated(
-    value = "by gRPC proto compiler (version 1.29.0)",
+    value = "by gRPC proto compiler (version 1.30.0)",
```
Why is that related?

kananindzya replied:
It's changed after running tests, I think. Now in the master branch the value is already "1.30.0"

eyalkoren commented on apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java:
```diff
@@ -78,5 +81,32 @@ public double get() {
             }
         });
+
+        List<MemoryPoolMXBean> memoryPoolMXBeans = ManagementFactory.getMemoryPoolMXBeans();
+
+        for (final MemoryPoolMXBean memoryPoolMXBean : memoryPoolMXBeans) {
+            if (memoryPoolMXBean.getType() != MemoryType.HEAP) {
+                continue;
+            }
+            final Labels memoryPoolTags = Labels.Mutable.of("type", memoryPoolMXBean.getName());
+            registry.add("jvm.memory.heap.used", memoryPoolTags, new DoubleSupplier() {
+                @Override
+                public double get() {
+                    return memoryPoolMXBean.getUsage().getUsed();
+                }
+            });
+            registry.add("jvm.memory.heap.committed", memoryPoolTags, new DoubleSupplier() {
+                @Override
+                public double get() {
+                    return memoryPoolMXBean.getUsage().getCommitted();
+                }
+            });
+            registry.add("jvm.memory.heap.max", memoryPoolTags, new DoubleSupplier() {
+                @Override
+                public double get() {
+                    return memoryPoolMXBean.getUsage().getMax();
+                }
+            });
+        }
```
As these metrics nice-to-have, please surround with try/catch, and either log error on failure or debug on success. It will help us test that we don't break anything on all JVMs we use in the integration tests (including J9)

kananindzya replied:
added try/catch

kananindzya commented:
> @kananindzya Looks great!
> A few asks:
> 
>     * Let's use `jvm.memory.heap.pool.XXX` for the metricset keys, instead of `jvm.memory.heap.XXX`
> 
>     * Please change the tag from `type` to `name`
> 
>     * Please add tests (and remove the system out)
> 
> 
> Thanks!

- changed to jvm.memory.heap.pool.XXX
- tag changed from "type" to "name"
- tests added + sout deleted :)
- added changelog

eyalkoren commented:
@kananindzya we only need to change the test, so it is not platform-dependent. Would you like to do that?

kananindzya commented:
> @kananindzya we only need to change the test, so it is not platform-dependent. Would you like to do that?

Hi, sorry for the delay. I unfortunately had no time last couple weeks. I am planning to do it this weekend.

eyalkoren commented:
No rush, I just wanted to know this PR is still alive 🙂

kananindzya commented:
> The test is dependent on JVM implementation. It should not assume specific memory pool names.

Fixed. I'm not sure if i have reached the goal :)

kananindzya commented on apm-agent-core/src/test/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetricsTest.java:
```diff
@@ -47,7 +53,23 @@ void testMetrics() {
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.max", spaceLabel)).isNotNaN();
+        }
         final long[] longs = new long[1000000];
         System.out.println(registry.toString());
```
removed
